### PR TITLE
Replace join over constant with project

### PR DESCRIPTION
**Pull request by kasiafi**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * This rule transforms plans with join where one of the sources is
+ * a single-row ValuesNode, and the join condition is `on true`.
+ * The JoinNode is replaced with the other join source and a projection
+ * which appends constant values from the ValuesNode.
+ * This rule is similar to ReplaceRedundantJoinWithSource.
+ * <p>
+ * Note 1: When transforming an outer join (LEFT, RIGHT or FULL), and an
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java:46]: If we had an "enforce at least one row" operation that produces nulls if there are no rows in the input, we could do this.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+        ValuesNode values = (ValuesNode) node;
+        if (values.getRowCount() != 1) {
+            return false;
+        }
+
+        if (values.getOutputSymbols().isEmpty()) {
+            // handled by ReplaceRedundantJoinWithSource rule
+            return false;
+        }
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java:157]: While convenient, this check doesn't belong in `isSingleConstantRow`. I hoist it outside of this method for clarity

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        if (isAtMost(node.getLeft(), context.getLookup(), 0) || isAtMost(node.getRight(), context.getLookup(), 0)) {
+            return Result.empty();
+        }
+
+        PlanNode left = context.getLookup().resolve(node.getLeft());
+        PlanNode right = context.getLookup().resolve(node.getRight());
+        boolean lefSourceSingleConstantRow = isSingleConstantRow(left);
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java:97]: Typo: lef

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+        Expression row = getOnlyElement(values.getRows().get());
+
+        return row instanceof Row;
+    }
+
+    private ProjectNode appendProjection(PlanNode source, List<Symbol> sourceOutputs, PlanNode constantSource, List<Symbol> constantOutputs, PlanNodeIdAllocator idAllocator)
+    {
+        Assignments.Builder assignments = Assignments.builder()
+                .putIdentities(sourceOutputs);
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java:167]: Move this down closer to where it's used. I would restructure this method as follows for better visual clarity:

```
        ValuesNode values = (ValuesNode) constantSource;
        Row row = (Row) getOnlyElement(values.getRows().get());

        Map<Symbol, Expression> mapping = new HashMap<>();
        for (int i = 0; i < values.getOutputSymbols().size(); i++) {
            mapping.put(values.getOutputSymbols().get(i), row.getItems().get(i));
        }

        Assignments.Builder assignments = Assignments.builder()
                .putIdentities(sourceOutputs);

        constantOutputs.stream()
                .forEach(symbol -> assignments.put(symbol, mapping.get(symbol)));

        return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
```

File: core/trino-main/src/test/java/io/trino/sql/planner/TestDereferencePushDown.java [hunk 5]
```diff
@@ -96,14 +104,22 @@ public void testDereferencePushdownJoin()
...
+        // dereference pushdown + constant folding
         assertPlan("WITH t(msg) AS (VALUES ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE))))" +
                         "SELECT a.msg.y, b.msg.x " +
                         "FROM t a CROSS JOIN t b " +
                         "WHERE a.msg.x = 7 OR IS_FINITE(b.msg.y)",
-                anyTree(
-                        join(INNER, ImmutableList.of(),
-                                values(ImmutableList.of("a_y"), ImmutableList.of(ImmutableList.of(new DoubleLiteral("2e0")))),
-                                values(ImmutableList.of("b_x"), ImmutableList.of(ImmutableList.of(new GenericLiteral("BIGINT", "1")))))));
+                any(project(
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestDereferencePushDown.java:112]: format as follows for clarity:

```
any(
    project(
        ...
```

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java (hunk 1 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java:46]: That would be similar to `EnforceSingleRowNode` / `EnforceSingleRowOperator`. The main difficulty when adding it is to make sure that the new node doesn't block any optimization chains (pruning, pushdown etc.).

**Pull request merged**